### ISSUE TITLE
WeightsHistHandler should plot all weights (incl those without grad)

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/ignite#2328]
## 🚀 Feature

Currently, `WeightsHistHandler` do not log weights without grad:
https://github.com/pytorch/ignite/blob/a168476762061f74af64793bce1b2d5b6e9b3f15/ignite/contrib/handlers/tensorboard_logger.py#L434-L435

Let's put an option to enable logging all weights. Maybe, we can make it even without option and just log everything ?

Meanwhile, here is a workaround code for TensorboardLogger:
```python
from ignite.contrib.handlers import TensorboardLogger
from ignite.contrib.handlers.tensorboard_logger import WeightsHistHandler

class FixedWeightsHistHandler(WeightsHistHandler):
    
    def __call__(self, engine, logger, event_name):
        if not isinstance(logger, TensorboardLogger):
            raise RuntimeError("Handler 'WeightsHistHandler' works only with TensorboardLogger")

        global_step = engine.state.get_event_attrib_value(event_name)
        tag_prefix = f"{self.tag}/" if self.tag else ""
        for name, p in self.model.named_parameters():            
            name = name.replace(".", "/")
            logger.writer.add_histogram(
                tag=f"{tag_prefix}weights/{name}", values=p.data.detach().cpu().numpy(), global_step=global_step,
            )

```

**Comment by DhDeepLIT** [comment]
Assume you have a partial frozen model, it makes sense to only trace the evolution of the weight which are updated.
The question is does checking if p.grad is not None is what we want ?
What about weights modification with other computation ? like pca_low_rank ?

**Comment by vfdev-5** [comment]
I see what you mean. Ultimately, we would like to have an optional list of weights to plot ?

**Comment by DhDeepLIT** [comment]
BTW it would be very nice to add a filter on the weights/grads to monitor : a list of str could be enough 

# keys_to_log : list of str 

for name, p in self.model.named_parameters():
            if p.grad is None:
                continue

            name = name.replace(".", "/")
            for k in keys_to_log:
                      if k in name:
                             logger.writer.add_histogram(
                                                 tag=f"{tag_prefix}grads/{name}", values=p.grad.detach().cpu().numpy(), global_step=global_step)
                            break

**Comment by vfdev-5** [comment]
@DhDeepLIT I thought you wanted to avoid the case of inexisting grads and log weights in any case ?
```python
            if p.grad is None:
               continue
```

**Comment by DhDeepLIT** [comment]
BTW your proposed patch is fine for my need, the whitelist is a great idea too.

**Comment by vfdev-5** [comment]
FYI: zero_grad does not set grads to None by default :)

https://github.com/pytorch/pytorch/blob/33e9a0b5f6674bd429bda689534e1c987b38cf6e/torch/nn/modules/module.py#L1785-L1794

**Comment by DhDeepLIT** [comment]
Maybe it should be the other way around though :
it is easier to filter the weights/grad we want if a key is present in the module full name say you want to check ALL fc layer. It would be better to check if 'fc' is in the named parameter of the module, right ?

**Comment by vfdev-5** [comment]
I agree that non-exact match for names_whitelist could be helpful...

**Comment by DhDeepLIT** [comment]
https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html

There is a option with 'set_to_none=True'

**Comment by sadra-barikbin** [comment]
Hi @vfdev-5 
I'm back. Assign it to me if you will.
